pkg/sql/authorization.go

```diff
@@ -414,85 +414,86 @@ func MemberOfWithAdminOption(
-		userMapping, found := func() (userRoleMembership, bool) {
-			roleMembersCache.Lock()
-			defer roleMembersCache.Unlock()
-			if roleMembersCache.tableVersion != tableVersion {
-				// Update version and drop the map.
-				roleMembersCache.tableVersion = tableVersion
-				roleMembersCache.userCache = make(map[security.SQLUsername]userRoleMembership)
-				roleMembersCache.boundAccount.Empty(ctx)
-			}
-			userMapping, ok := roleMembersCache.userCache[member]
-			return userMapping, ok
-		}()
-
-		if found {
-			// Found: return.
-			return userMapping, nil
+	// Check version and maybe clear cache while holding the mutex.
+	// We use a closure here so that we release the lock here, then keep
+	// going and re-lock if adding the looked-up entry.
+	userMapping, found := func() (userRoleMembership, bool) {
+		roleMembersCache.Lock()
+		defer roleMembersCache.Unlock()
+		if roleMembersCache.tableVersion < tableVersion {
+			// If the cache is based on an old table version, then update version and
+			// drop the map.
+			roleMembersCache.tableVersion = tableVersion
+			roleMembersCache.userCache = make(map[security.SQLUsername]userRoleMembership)
+			roleMembersCache.boundAccount.Empty(ctx)
+		} else if roleMembersCache.tableVersion > tableVersion {
+			// If the cache is based on a newer table version, then this transaction
+			// should not use the cached data.
+			return nil, false
 		}
-
-		// Lookup memberships outside the lock, with at most one request in-flight
-		// for each user The role_memberships table versions is also part
-		// of the request key so that we don't read data from an old version
-		// of the table.
-		ch, _ := roleMembersCache.populateCacheGroup.DoChan(
-			fmt.Sprintf("%s-%d", member.Normalized(), tableVersion),
-			func() (interface{}, error) {
-				// Use a different context to fetch, so that it isn't possible for
-				// one query to timeout and cause all the goroutines that are waiting
-				// to get a timeout error.
-				ctx, cancel := roleMembersCache.stopper.WithCancelOnQuiesce(
-					logtags.WithTags(context.Background(), logtags.FromContext(ctx)))
-				defer cancel()
-				return resolveMemberOfWithAdminOption(ctx, member, ie, txn, useSingleQueryForRoleMembershipCache.Get(execCfg.SV()))
-			},
-		)
-		var memberships map[security.SQLUsername]bool
-		select {
-		case res := <-ch:
-			if res.Err != nil {
-				return nil, res.Err
-			}
-			memberships = res.Val.(map[security.SQLUsername]bool)
-		case <-ctx.Done():
-			return nil, ctx.Err()
+		userMapping, ok := roleMembersCache.userCache[member]
+		return userMapping, ok
+	}()
+
+	if found {
+		// Found: return.
+		return userMapping, nil
+	}
+
+	// Lookup memberships outside the lock. There will be at most one request
+	// in-flight for each user. The role_memberships table version is also part
+	// of the request key so that we don't read data from an old version of the
+	// table.
+	ch, _ := roleMembersCache.populateCacheGroup.DoChan(
```

Comment:
do we need the `, _` in `ch, _` here?

Reply:
no, it just returns a bool indicating if this call was the one that initiated a request, or if this call delegated to an existing request that was in flight. it's not important for this

https://github.com/cockroachdb/cockroach/blob/4d987b3f0788c98fa358715b336a1c55e254ee50/pkg/util/syncutil/singleflight/singleflight.go#L86-L88

Reply:
oh, sorry, i was reading the golang docs for this